src/err/mod.rs

```diff
@@ -1070,17 +1086,76 @@ mod tests {
-            let mut fields = debug_str["PyErr { ".len()..debug_str.len() - 2].split(", ");
+            // Strip "PyErr { " and " }". Split into 3 substrings to separate type,
+            // value, and traceback while not splitting the string within traceback.
+            let mut fields = debug_str["PyErr { ".len()..debug_str.len() - 2].splitn(3, ", ");
 
             assert_eq!(fields.next().unwrap(), "type: <class 'Exception'>");
             assert_eq!(fields.next().unwrap(), "value: Exception('banana')");
+            assert_eq!(
+                fields.next().unwrap(),
+                "traceback: Some(\"Traceback (most recent call last):\\n  File \\\"<string>\\\", line 1, in <module>\\n\")"
+            );
 
-            let traceback = fields.next().unwrap();
-            assert!(traceback.starts_with("traceback: Some(<traceback object at 0x"));
-            assert!(traceback.ends_with(">)"));
+            assert!(fields.next().is_none());
+        });
+    }
+
+    #[test]
+    fn err_debug_unformattable() {
+        // Debug representation should be like the following (without the newlines):
+        // PyErr {
+        //     type: <class 'Exception'>,
+        //     value: Exception('banana'),
+        //     traceback: Some(\"<unformattable <traceback object at 0x...>>\")
+        // }
+
+        Python::with_gil(|py| {
+            let err = py
+                .run(
+                    ffi::c_str!(
+                        r#"
+import io, sys, unittest.mock
+sys.modules['orig_io'] = sys.modules['io']
+sys.modules['io'] = unittest.mock.Mock()
+raise Exception('banana')"#
+                    ),
+                    None,
+                    None,
+                )
+                .expect_err("raising should have given us an error");
+
+            let debug_str = format!("{:?}", err);
+            assert!(debug_str.starts_with("PyErr { "));
+            assert!(debug_str.ends_with(" }"));
 
+            // Strip "PyErr { " and " }". Split into 3 substrings to separate type,
+            // value, and traceback while not splitting the string within traceback.
+            let mut fields = debug_str["PyErr { ".len()..debug_str.len() - 2].splitn(3, ", ");
+
+            assert_eq!(fields.next().unwrap(), "type: <class 'Exception'>");
+            assert_eq!(fields.next().unwrap(), "value: Exception('banana')");
+
+            let traceback = fields.next().unwrap();
+            assert!(
+                traceback.starts_with("traceback: Some(\"<unformattable <traceback object at 0x"),
+                "assertion failed, actual traceback str: {:?}",
+                traceback
+            );
             assert!(fields.next().is_none());
+
+            py.run(
+                ffi::c_str!(
+                    r#"
+import io, sys, unittest.mock
+sys.modules['io'] = sys.modules['orig_io']
+del sys.modules['orig_io']
+"#
+                ),
+                None,
+                None,
+            )
+            .unwrap();
```

Comment:
If we go down this route, this probably needs to be implemented via a `drop` guard. I think all unit tests run within the same interpreter, so if any assertion here fails, this will not get called and other following tests may fail due to this modification.

On the free-threaded build the whole approach might be problematic because the tests can actually run in parallel. (maybe @ngoldbaum has some comments on that)

It would probably be better to use an integration test for this, so that it gets its own process.

Reply:
Yes, please avoid using global state in cargo unit tests - they all run in parallel in the free-threaded build.

On the other hand pytest serializes test execution (unless you explicitly spawn threads or use a pytest plugin that runs tests in parallel) so it is OK to use global state in a test written in Python.

Reply:
Moved the test to an integration test.